Exception that actives with a massive number of Bluetooth signals (beacon and smatphone)

Tested with more devices (samsung smartphone and tablet with OS 4.4.2 and 5) when start scan, the app, after few seconds, crashes with this error:

java.lang.RuntimeException: An error occured while executing doInBackground()
	at android.os.AsyncTask$3.done(AsyncTask.java:300)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:355)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:222)
	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	at java.lang.Thread.run(Thread.java:818)
Caused by: java.lang.ArrayIndexOutOfBoundsException: length=62; index=62
	at org.altbeacon.beacon.BeaconParser.byteArraysMatch(BeaconParser.java:777)
	at org.altbeacon.beacon.BeaconParser.fromScanData(BeaconParser.java:416)
	at org.altbeacon.beacon.BeaconParser.fromScanData(BeaconParser.java:373)
	at org.altbeacon.beacon.service.BeaconService$ScanProcessor.doInBackground(BeaconService.java:457)
	at org.altbeacon.beacon.service.BeaconService$ScanProcessor.doInBackground(BeaconService.java:442)
	at android.os.AsyncTask$2.call(AsyncTask.java:288)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)

The stack trace indicates that the library crashed trying to process malformed Bluetooth LE packet.   There is a flaw in line BeaconParser.java:777 in the byteArraysMatch method that overflows an array in certain cases.

```
Caused by: javalangArrayIndexOutOfBoundsException: length=62; index=62
at orgaltbeaconbeaconBeaconParserbyteArraysMatch(BeaconParserjava:777)
at org.altbeacon.beacon.BeaconParser.fromScanData(BeaconParser.java:416)
```

--- a/src/main/java/org/altbeacon/beacon/BeaconParser.java
+++ b/src/main/java/org/altbeacon/beacon/BeaconParser.java
@@ -773,6 +773,9 @@ public class BeaconParser {
 
     private boolean byteArraysMatch(byte[] array1, int offset1, byte[] array2, int offset2) {
         int minSize = array1.length > array2.length ? array2.length : array1.length;
+        if (offset1+minSize > array1.length || offset2+minSize > array2.length) {
+            return false;
+        }
         for (int i = 0; i <  minSize; i++) {
             if (array1[i+offset1] != array2[i+offset2]) {
                 return false;

--- a/src/test/java/org/altbeacon/beacon/BeaconParserTest.java
+++ b/src/test/java/org/altbeacon/beacon/BeaconParserTest.java
@@ -290,5 +290,20 @@ public class BeaconParserTest {
         byte[] bytes = p.getBeaconAdvertisementData(beacon);
         assertEquals("First byte of url should be in position 3", 0x02, bytes[2]);
     }
+    @Test
+    public void doesNotCashWithOverflowingByteCodeComparisonOnPdu() {
+        // Test for https://github.com/AltBeacon/android-beacon-library/issues/323
+        org.robolectric.shadows.ShadowLog.stream = System.err;
+
+        // Note that the length field below is 0x16 instead of 0x1b, indicating that the packet ends
+        // one byte before the second identifier field starts
+
+        byte[] bytes = hexStringToByteArray("02010604ffe000be");
+        BeaconParser parser = new BeaconParser();
+        parser.setBeaconLayout("m:2-3=beac,i:4-19,i:20-21,i:22-23,p:24-24,d:25-25");
+
+        Beacon beacon = parser.fromScanData(bytes, -55, null);
+        assertNull("beacon not be parsed without an exception being thrown", beacon);
+    }
 
 }

